logging/logging.go

```diff
@@ -923,3 +928,41 @@ func printEntryToWriter(entry *logpb.LogEntry, w io.Writer) error {
 	}
 	return err
 }
+
+func (l *Logger) redirectAsJSON(pbe *logpb.LogEntry) (err error) {
+	err = serializeEntryToWriter(pbe, l.redirectOutputWriter)
+	if err == nil {
+		if internal.IngestInstrumentation {
+			internal.IngestInstrumentation = false
+			pbe, err = l.instrumentationEntry()
+			if err == nil {
+				err = serializeEntryToWriter(pbe, l.redirectOutputWriter)
+			}
+		}
+	}
+	return
+}
+
+func (l *Logger) ingestInstrumentation(entries []*logpb.LogEntry) ([]*logpb.LogEntry, bool) {
+	partialSuccess := false
+	if internal.IngestInstrumentation {
+		internal.IngestInstrumentation = false
+		instrumentation, err := l.instrumentationEntry()
+		if err == nil {
+			partialSuccess = true // ensure instrumentation will be ingested even if ent is too big or invalid
+			instrumentation.LogName = internal.LogPath(l.client.parent, "diagnostic-log")
+			entries = append(entries, instrumentation)
+		}
+	}
+	return entries, partialSuccess
+}
+
+func (l *Logger) instrumentationEntry() (*logpb.LogEntry, error) {
+	ent := Entry{
+		Payload: map[string]*telemetryPayload{
```

Comment:
This whole map could be cached.

Reply:
According to the current design this is sent only once. Although the object is small, keeping it in memory is not required.